lib/async_hooks.js

```diff
@@ -223,31 +223,24 @@ const storageHook = createHook({
 class AsyncLocalStorage {
   constructor() {
     this.kResourceStore = Symbol('kResourceStore');
-    this.enabled = false;
   }
 
   disable() {
-    if (this.enabled) {
-      this.enabled = false;
-      // If this.enabled, the instance must be in storageList
-      storageList.splice(storageList.indexOf(this), 1);
-      if (storageList.length === 0) {
-        storageHook.disable();
-      }
+    this._exit();
```

Comment:
This new approach without `enabled` property fails in the following test, while the previous implementation passes it:
```js
'use strict';
require('../common');
const assert = require('assert');
const { AsyncLocalStorage } = require('async_hooks');

const asyncLocalStorage = new AsyncLocalStorage();

asyncLocalStorage.runSyncAndReturn({}, () => {
  process.nextTick(() => {
    assert.strictEqual(asyncLocalStorage.getStore(), undefined);
  });
  asyncLocalStorage.disable();
});
```

Reply:
I'm not sure the _original_ `disable()` behaviour is really _correct_ either though, given that concurrent `run(...)` calls will interfere with the enabled state. For example, consider that code above, but replace `process.nextTick(...)` with something with less deterministic sequencing in the event loop, say `fs.readFile(...)`. What happens if _another_ request starts a `run(...)` _after_ the tick that called `disable()`, but _before_ the http request callback runs? The entire `AsyncLocalStorage` will now be set back to enabled, and the first request will continue as if it was never disabled.

Reply:
That's the intended behavior. `disable()` is here to allow the instance being garbage collected when not used anymore. In this example, it is still used and has to be re-enabled.

Reply:
I restored the previous logic, so now this PR only fixes the context leak and merges the methods.